2024_wsb/Src/halleffectsensor.c

```diff
@@ -0,0 +1,109 @@
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+
+#include "bsp.h"
+#include "cmsis_os.h"
+#include "debug.h"
+#include "detectWSB.h"
+#include "main.h"
+#include "multiSensorADC.h"
+#include "task.h"
+#include "tim.h"
+#if BOARD_ID == ID_WSBRL
+#include "wsbrl_can.h"
+#elif BOARD_ID == ID_WSBRR
+#include "wsbrr_can.h"
+#endif
+
+#define NUM_TEETH 16
+#define HALL_EFFECT_TASK_PERIOD 100
+#define WHEEL_RADIUS 0.40005
+#define PI 3.14156
+#define TICKS_PER_SECOND 1000
+
+/*
+ * IOC FILE CHANGES:
+ * PE13:
+ *  state: GPIO_EXTI13
+ *  GPIO MODE: external interrupt mode w/ pull-up resistor
+ *  user label: REAR_HALL_EFFECT_ENCODER
+ *  EXTI[15:10] interrupts: enabled
+ *
+ * TIM5:
+ *  clock source: internal clock
+ *  prescaler: 39999
+ *  auto preload reload: enabled
+ *  note: clock runs at 40 MHz
+ */
+
+
+float getRps(uint32_t count, uint32_t tick_diff) {
+    /*
+     * teeth/s = count diff/time diff
+     * rps = 1/num_teeth * teeth/s
+     */
+    return (1.0f*count) / (1.0f*tick_diff/TICKS_PER_SECOND) / NUM_TEETH;
+}
+
+int64_t getRpm(float rps) {
+    /*
+     * rpm = 60rps
+     * 2024 car: 16 teeth
+     */
+    return (int64_t)(int32_t)(rps * 60.0f);
+}
+
+float getKph(float rps) {
+    /*
+     * rad/s = 2pi * rps
+     * m/s = rad/s * RADIUS
+     * kph = 3.6 * m/s
+     * 2024 car: effective radius 0.40005m
+     */
+    return rps * (2.0f*PI) * WHEEL_RADIUS * 3.6f;
+}
+
+uint32_t pulse_count = 0;
+void HAL_GPIO_EXTI_Callback(uint16_t GPIO_Pin) {
```

Comment:
Having this as an interrupt will slow down the program. Say the wheel is spinning really fast ~100rps. This would cause the entire program to be halted 100 times a second just to increment a variable. Check out this for how to enable external trigger pulses on a timer without an interrupt: https://community.st.com/t5/stm32-mcus-products/count-external-pulses-using-timer/td-p/59349